middlewares/prometheus_test.go

```diff
@@ -42,6 +44,81 @@ func TestPrometheus(t *testing.T) {
 		t.Errorf("body does not contain request total entry '%s'", reqsName)
 	}
 	if !strings.Contains(body, latencyName) {
-		t.Errorf("body does not contain request duration entry '%s'", reqsName)
+		t.Errorf("body does not contain request duration entry '%s'", latencyName)
+	}
+
+	// Register the same metrics again
+	metricsMiddlewareBackend = NewMetricsWrapper(NewPrometheus("test", &types.Prometheus{}))
+	n = negroni.New()
+	n.Use(metricsMiddlewareBackend)
+	n.UseHandler(r)
+
+	n.ServeHTTP(recorder, req2)
+
+	metricsFamily, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		t.Error(err)
+	}
+
+	var latencyMetricFamily, reqsMetricFamily *dto.MetricFamily
+
+	for _, metricFamily := range metricsFamily {
+		if metricFamily.GetName() == reqsName {
+			reqsMetricFamily = metricFamily
+		}
+		if metricFamily.GetName() == latencyName {
+			latencyMetricFamily = metricFamily
+		}
+	}
+
+	if reqsMetricFamily == nil {
```

Comment:
The test in its current implementation has two shortcomings:

- If `reqsMetricFamily` here (or `latencyMetricFamily` later) is `nil`, we still continue with the test execution. At best, this leads to follow-up test errors that we should not have to care about but will aggravate interpreting the test result. At worst, we run into nil pointer dereferences in lines 106, 117, and/or 121.
- We cannot easily ignore/shortcut an individual `nil` case given the chosen test design.

There are a few options to fix this. Here's the one I prefer:

- Rename the `expectedLabelsForMetricFamily` struct into `tests` and have it include a custom `assert` function of type `func(*dto.MetricFamily)`. This function is supposed to implement the check in line 117 for the `reqsMetricFamily` case and the one in line 121 for `latencyMetricFamily`. (The idea is somewhat similar to the `parserFunc` in [this test](https://github.com/krancour/traefik/blob/47bfbb5ebcf664022200287f01873088aeadf4a1/middlewares/authenticator_test.go#L16).)
- Move the lookup of the metric families by metric name into the test loop (i.e., between line 105 and 106). This way, we can `t.Errorf` and `continue` the loop if we cannot find a metric family. Also, we can move the `metricFamily` variable from the test struct to the stack inside the loop.
- Execute `assert` in the loop, passing it the metric family we just looked up.

I'd also suggest to move the lookup functionality into a small helper function, i.e., `findMetricFamily(name string, families []*dto.MetricFamily) *dto.MetricFamily`.